### ISSUE TITLE
End game when no move for king is checked

### ISSUE THREAD

**Issue by kimnamcham** [emdio/kitteneitor#8]
First, thanks a lot for you engine, i'm using it for my game, it's so wonderful and easy to use. But it's have some problems. When king is checked, no move to avoid ,engine has no method to check it. I saw the method isEndGame but it's so simply to check number of chess on board. So can you give me some solutions or suggest , thanks :)

**Comment by emdio** [comment]
I think there's some kind of confussion here; isEndGame function checks whether we are in the endgame, the final stage of the game (few pieces, no queens, etc), and not whether the game is finished.
About avoiding a check; Kitteneitor has a pseudo-legal move generator; this means that in a certain position, no matter it's in check or not, Kitt will create a list of all the possible moves; then Kitt will play each move on the board, and after doing the move, it'll check whether it's in check or not; if it's in check, it'll just undo the move and will take the next move in the list. If after taking all the moves in the list, none of them gives a position in which Kitt is not in check, then it's checkmatted.
Not sure if this answers your questions; if it's not the case, don't hesitate to ask again.

**Comment by kimnamcham** [comment]
I don't understand what did you said .i'm using your engine for my game on ui, when player click and move piece, it's genarate postion move. After that, i will call computer chess to genarate move and call back number move . How can i check when game is over (checkmate) for player and computer chess ? I search google and find a solution : genarate all move for player( or computer) and check it's in checked or not, if it's in checked and no move gen -> checkmate. But i don't know how to do that :(

**Comment by emdio** [comment]
"genarate all move for player( or computer) and check it's in checked or not, if it's in checked and no move gen -> checkmate. But i don't know how to do that :(" That's what I tried to explain in my first message.

Check the code in search.c after the comment 
/* Once we've checked all the moves, if we have no legal moves,
* then that's checkmate or stalemate */

On the other hand, maybe you can use Kitteneitor's output score. If its last move has checkmated the opponent, the score should be 9999. If the last move made by the opponent has checkmated Kitt, the the score should be -10000.

**Comment by kimnamcham** [comment]
Thanks you very much, i used engine output to checkmate :D :+1:

**Comment by emdio** [comment]
Any time.

BTW, there's a bug in Kitt's detection of three moves repetition; it just doesn't work properly. This bug doesn't crash the engine or anything like that, but once in a while it'll allow an easy draw for the opponent. Other than that, Kitt has played thousands of games in FICS with no issues:
http://ficsgames.org/cgi-bin/search.cgi?action=statistics;player=Kitteneitor

**Comment by kimnamcham** [comment]
I agree with you, sometimes it's crash at memcpy when i load 1 book on 300 book of Lev Alburt. If i set maxdepth =6 and force . How can i replace or modifiy your search algorithm by other engine as stockfish,...

**Comment by emdio** [comment]
That's interesting. Could you open a new issue with that 1 position from the Lev Alburt suite?

About replacing Kitt's search with the one from another engine, I'm afraid it's not that easy; of course you can get lots of ideas, but implementing these ideas in Kitt's code will need of your typing it on code.

**Comment by kimnamcham** [comment]
I modified function test1 and use command "on", here is 1 position from the Lev Alburt:
	int piece_test[64]={
		EMPTY, KNIGHT, EMPTY, QUEEN, EMPTY, ROOK, KING, EMPTY,
		KNIGHT, BISHOP, EMPTY, EMPTY, PAWN, PAWN, BISHOP, PAWN,
		PAWN, PAWN, EMPTY, EMPTY, EMPTY, EMPTY, PAWN, EMPTY,
		EMPTY, EMPTY, EMPTY, PAWN, EMPTY, EMPTY, EMPTY, EMPTY,
		EMPTY, EMPTY, EMPTY, PAWN, KNIGHT, EMPTY, EMPTY, EMPTY,
		BISHOP, PAWN, EMPTY, BISHOP, PAWN, KNIGHT, EMPTY, EMPTY,
		PAWN, EMPTY, EMPTY, EMPTY, QUEEN, PAWN, PAWN, PAWN,
		EMPTY, EMPTY, ROOK, EMPTY, EMPTY, EMPTY, KING, EMPTY };
	int color_test[64]={
		EMPTY, BLACK, EMPTY, BLACK, EMPTY, BLACK, BLACK, EMPTY,
		WHITE, BLACK, EMPTY, EMPTY, BLACK, BLACK, BLACK, BLACK,
		BLACK, BLACK, EMPTY, EMPTY, EMPTY, EMPTY, BLACK, EMPTY,
		EMPTY, EMPTY, EMPTY, BLACK, EMPTY, EMPTY, EMPTY, EMPTY,
		EMPTY, EMPTY, EMPTY, WHITE, BLACK, EMPTY, EMPTY, EMPTY,
		WHITE, WHITE, EMPTY, WHITE, WHITE, WHITE, EMPTY, EMPTY,
		WHITE, EMPTY, EMPTY, EMPTY, WHITE, WHITE, WHITE, WHITE,
		EMPTY, EMPTY, WHITE, EMPTY, EMPTY, EMPTY, WHITE, EMPTY};
